shaders.go

```diff
@@ -278,16 +292,489 @@ func setBufferValue(buffer []float32, index int, value float32, changed *bool) {
 	}
 }
 
+type legacyShader struct {
+	program *gl.Program
+
+	indicesRectangles    []uint16
+	indicesRectanglesVBO *gl.Buffer
+
+	inPosition int
+	inColor    int
+
+	matrixProjection *gl.UniformLocation
+	matrixView       *gl.UniformLocation
+	matrixModel      *gl.UniformLocation
+	inRadius         *gl.UniformLocation
+	inCenter         *gl.UniformLocation
+	inViewport       *gl.UniformLocation
+	inBorderWidth    *gl.UniformLocation
+	inBorderColor    *gl.UniformLocation
+
+	projectionMatrix []float32
+	viewMatrix       []float32
+	modelMatrix      []float32
+
+	cameraEnabled bool
+
+	lastBuffer *gl.Buffer
+}
+
+func (l *legacyShader) Initialize() {
+	l.program = LoadShader(`
+attribute vec2 in_Position;
+attribute vec4 in_Color;
+
+uniform mat3 matrixProjection;
+uniform mat3 matrixView;
+uniform mat3 matrixModel;
+uniform vec2 in_Radius;
+uniform vec2 in_Center;
+uniform vec2 in_Viewport;
+uniform float in_BorderWidth;
+uniform vec4 in_BorderColor;
+
+varying vec4 var_Color;
+varying vec2 var_Radius;
+varying vec2 var_Center;
+varying float var_BorderWidth;
+varying vec4 var_BorderColor;
+
+void main() {
+  var_Color = in_Color;
+
+  vec3 matr = matrixProjection * matrixView * matrixModel * vec3(in_Position, 1.0);
+  gl_Position = vec4(matr.xy, 0, matr.z);
+
+  if (in_Radius.x > 0.0 && in_Radius.y > 0.0)
+  {
+    var_Radius = in_Radius;
+    var_BorderWidth = in_BorderWidth;
+    var_BorderColor = in_BorderColor;
+
+    vec3 vecCenter = (matrixProjection * matrixView * matrixModel * vec3(in_Center, 1.0));
+    var_Center = (vecCenter.xy/vecCenter.z + vec2(1.0, 1.0)) * in_Viewport / vec2(2.0);
+  } else {
+    var_Radius = vec2(0.0, 0.0);
+  }
+}
+`, `
+#ifdef GL_ES
+#define LOWP lowp
+precision mediump float;
+#else
+#define LOWP
+#endif
+
+varying vec4 var_Color;
+varying vec2 var_Radius;
+varying vec2 var_Center;
+varying float var_BorderWidth;
+varying vec4 var_BorderColor;
+
+void main (void) {
+  gl_FragColor = var_Color;
+
+  float halfBorder = var_BorderWidth / 2;
+
+  if (var_Radius.x > 0.0 && var_Radius.y > 0.0)
+  {
+    if (pow(gl_FragCoord.x - var_Center.x, 2.0) / pow(var_Radius.x - halfBorder, 2.0) + pow(gl_FragCoord.y - var_Center.y, 2.0) / pow(var_Radius.y - halfBorder, 2.0) > 1.0)
```

Comment:
This line basically says: if this single pixel we're looking at, is outside of the circle (by looking at the range from the center of the circle), then ...